packages/contracts/contracts/voting/multisig/Multisig.sol

```diff
@@ -0,0 +1,398 @@
+    function initialize(
+        IDAO _dao,
+        uint256 _minApprovals,
+        address[] calldata _members
+    ) public initializer {
+        __PluginUUPSUpgradeable_init(_dao);
+
+        // add member addresses to the address list
+        _updateAddresslist(_members, true);
+        _updateMinApprovals(_minApprovals);
+
+        emit AddressesAdded({members: _members});
+        emit MinApprovalUpdated({minApprovals: _minApprovals});
+    }
+
+    /// @notice Checks if this or the parent contract supports an interface by its ID.
+    /// @param _interfaceId The ID of the interface.
+    /// @return bool Returns `true` if the interface is supported.
+    function supportsInterface(bytes4 _interfaceId)
+        public
+        view
+        virtual
+        override(ERC165Upgradeable, GovernancePluginUUPSUpgradeable)
+        returns (bool)
+    {
+        return _interfaceId == MULTISIG_INTERFACE_ID || super.supportsInterface(_interfaceId);
+    }
+
+    /// @notice Returns the support threshold parameter stored in the voting settings.
+    /// @return The support threshold parameter.
+    function minApprovals() public view virtual returns (uint256) {
+        return minApprovals_;
+    }
+
+    /// @notice Returns the number of approvals,
+    /// @param _proposalId The ID of the proposal.
+    /// @return The number of approvals.
+    function approvals(uint256 _proposalId) public view returns (uint256) {
+        return proposals[_proposalId].tally.approvals;
+    }
+
+    /// @notice Updates the minimal approval parameter.
+    /// @param _minApprovals The new minimal approval value.
+    function updateMinApprovals(uint256 _minApprovals)
+        external
+        auth(UPDATE_MULTISIG_SETTINGS_PERMISSION_ID)
+    {
+        _updateMinApprovals(_minApprovals);
+
+        emit MinApprovalUpdated({minApprovals: _minApprovals});
+    }
+
+    /// @notice Adds new members to the address list and updates the minimum approval parameter.
+    /// @param _members The addresses of the members to be added.
+    function addAddresses(address[] calldata _members)
+        external
+        auth(UPDATE_MULTISIG_SETTINGS_PERMISSION_ID)
+    {
+        _updateAddresslist(_members, true);
+
+        emit AddressesAdded({members: _members});
+    }
+
+    /// @notice Removes existing members from the address list. Previously, it checks if the new addresslist length at least as long as the minimum approvals parameter requires. Note that `minApprovals` is must be at least 1 so the address list cannot become empty.
+    /// @param _members The addresses of the members to be removed.
+    function removeAddresses(address[] calldata _members)
+        external
+        auth(UPDATE_MULTISIG_SETTINGS_PERMISSION_ID)
+    {
+        // Check if the new addresslist has become shorter than the current minimum number of approvals required.
+        uint256 newAddresslistLength = addresslistLength(0) - _members.length;
```

Comment:
Here we are assuming that all array members were already authorized. But it would be possible that we had a 3 address multisig, we passed a list of 5 addresses to remove, none of which was part of the list. 

We would be failing because of that, whereas the changes would have no effect.

Also, why `addresslistLength(0)`? Does it mean at checkpoint zero?

Reply:
> Also, why addresslistLength(0)? Does it mean at checkpoint zero?
`/// @param _blockNumber The specific block to get the count from. If `0`, then the latest checkpoint value is returned.`

This is obviously unintuitive so I improved it in 366a2f74e0071fb3c4fdf04de49eca781b63b5bc

Reply:
I was under the impression that the call would revert if the member is not present because we already used this in AddresslistVoting.

You just discovered a **very** serious bug that was already present in `AddresslistVoting` and that the auditors overlooked. It allows you to increase the `addresslistLength` to whatever value you like so that the `minParticipation` criterium is never met.

I added the following test to `AddresslistVoting` and it failed 😳
```js
it('should NOT increase the address list length when adding a member multiple times', async () => {
      await voting.initialize(dao.address, votingSettings, addresslist(0));
      
      // add signer[0]
      await voting.addAddresses([signers[0].address]);
      await ethers.provider.send('evm_mine', []);
      expect(await voting.addresslistLength()).to.equal(1); // <-- 

      // add signer[0] again
      await voting.addAddresses([signers[0].address]);
      await ethers.provider.send('evm_mine', []);
      expect(await voting.addresslistLength()).to.equal(1); // <-- this fails because addreslistLength is now 2 
    });
```

Reply:
Fixed and thoroughly tested in ee83dd7f202c806f3342bad0f87a5f51ee057570